hail/src/test/scala/is/hail/expr/ir/IRSuite.scala

```diff
@@ -2999,4 +3000,40 @@ class IRSuite extends HailSuite {
             Row(Locus("20", 10277621)), Row(Locus("20", 11898992)), includesStart = true, includesEnd = false)),
         v))
   }
+
+  @Test def testSimpleTailLoop(): Unit = {
+    val triangleSum: IR = TailLoop("f",
+      FastIndexedSeq("x" -> In(0, TInt32()), "accum" -> I32(0)),
+      If(Ref("x", TInt32()) <= I32(0),
+        Ref("accum", TInt32()),
+        Recur("f",
+          FastIndexedSeq(
+            Ref("x", TInt32()) - I32(1),
+            Ref("accum", TInt32()) + Ref("x", TInt32())),
+          TInt32())))
+
+    assertEvalsTo(triangleSum, FastIndexedSeq(5 -> TInt32()), 15)
+  }
+
+  @Test def testNestedTailLoop(): Unit = {
+    val triangleSum: IR = TailLoop("f1",
+      FastIndexedSeq("x" -> In(0, TInt32()), "accum" -> I32(0)),
+      If(Ref("x", TInt32()) <= I32(0),
+        TailLoop("f2",
+          FastIndexedSeq("x2" -> Ref("accum", TInt32()), "accum2" -> I32(0)),
+          If(Ref("x2", TInt32()) <= I32(0),
+            Ref("accum2", TInt32()),
+            Recur("f2",
+              FastIndexedSeq(
+                Ref("x2", TInt32()) - I32(5),
+                Ref("accum2", TInt32()) + Ref("x2", TInt32())),
+              TInt32()))),
+        Recur("f1",
+          FastIndexedSeq(
+            Ref("x", TInt32()) - I32(1),
+            Ref("accum", TInt32()) + Ref("x", TInt32())),
+          TInt32())))
+
+    assertEvalsTo(triangleSum, FastIndexedSeq(5 -> TInt32()), 15 + 10 + 5)
+  }
```

Comment:
can we add a test of the semantics when various pieces of the loop return missing?

Args, accumulators, condition, etc.